settings/request.go

```diff
@@ -0,0 +1,278 @@
+//	|                            | result set metadata     |         |                           |
+//	|                            | will have more than     |         |                           |
+//	|                            | just name and type.     |         |                           |
+//	+----------------------------+-------------------------+---------+---------------------------+
+//	|       SQLParserDebug       | Defines whether to show |  false  | Since 2.3.1 (only if      |
+//	|                            | parser steps for        |         | built with                |
+//	|                            | following statements.   |         | -DCMAKE_BUILD_TYPE=Debug) |
+//	+----------------------------+-------------------------+---------+---------------------------+
+//	|    SQLRecursiveTriggers    | Defines whether a       |  true   | Since 2.3.1.              |
+//	|                            | triggered statement can |         |                           |
+//	|                            | activate a trigger.     |         |                           |
+//	+----------------------------+-------------------------+---------+---------------------------+
+//	| SQLReverseUnorderedSelects | Defines defines whether |  false  | Since 2.3.1.              |
+//	|                            | result rows are usually |         |                           |
+//	|                            | in reverse order if     |         |                           |
+//	|                            | there is no ORDER BY    |         |                           |
+//	|                            | clause.                 |         |                           |
+//	+----------------------------+-------------------------+---------+---------------------------+
+//	|       SQLSelectDebug       | Defines whether to show |  false  | Since 2.3.1 (only if      |
+//	|                            | to show execution steps |         | built with                |
+//	|                            | during SELECT.          |         | -DCMAKE_BUILD_TYPE=Debug) |
+//	+----------------------------+-------------------------+---------+---------------------------+
+//	|        SQLVDBEDebug        | Defines whether VDBE    |  false  | Since 2.3.1 (only if      |
+//	|                            | debug mode is enabled.  |         | built with                |
+//	|                            |                         |         | -DCMAKE_BUILD_TYPE=Debug) |
+//	+----------------------------+-------------------------+---------+---------------------------+
+//
+// Since: 1.10.0
+//
+// See also:
+//
+// * Session settings https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/_session_settings/
+package settings
+
+import (
+	"context"
+
+	"github.com/tarantool/go-tarantool"
+)
+
+// Request helps to work with session settings.
+type Request struct {
+	get *tarantool.SelectRequest
+	set *tarantool.UpdateRequest
+}
```

Comment:
Maybe this will help get rid of the checks in the code below: ```if req.set != nil {`
```Go
type contextRequest interface {
     tarantool.Request
     Context(ctx context.Context)
}

type Request struct {
    req contextRequest
}
```

Reply:
It looks easier:

```patch
diff --git a/settings/request.go b/settings/request.go
index fa045db..08be31d 100644
--- a/settings/request.go
+++ b/settings/request.go
@@ -65,23 +65,13 @@ import (
 
 // Request helps to work with session settings.
 type Request struct {
-    get *tarantool.SelectRequest
-    set *tarantool.UpdateRequest
+    req tarantool.Request
 }
 
-func (req *Request) assertBuiltWithNew() {
-    if (req.get == nil) && (req.set == nil) {
-        panic("Invalid settings.Request (empty), possibly build not with New*Request function")
-    }
-
-    if (req.get != nil) && (req.set != nil) {
-        panic("Invalid settings.Request (both types), possibly build not with New*Request function")
-    }
-}
 
 func newSetRequest(setting string, value interface{}) *Request {
     return &Request{
-        set: tarantool.NewUpdateRequest(sessionSettingsSpace).
+        req: tarantool.NewUpdateRequest(sessionSettingsSpace).
             Key(tarantool.StringKey{S: setting}).
             Operations(tarantool.NewOperations().Assign(sessionSettingValueField, value)),
     }
@@ -89,7 +79,7 @@ func newSetRequest(setting string, value interface{}) *Request {
 
 func newGetRequest(setting string) *Request {
     return &Request{
-        get: tarantool.NewSelectRequest(sessionSettingsSpace).
+        req: tarantool.NewSelectRequest(sessionSettingsSpace).
             Key(tarantool.StringKey{S: setting}).
             Limit(1),
     }
@@ -97,12 +87,11 @@ func newGetRequest(setting string) *Request {
 
 // Context sets a passed context to the session settings request.
 func (req *Request) Context(ctx context.Context) *Request {
-    req.assertBuiltWithNew()
-
-    if req.set != nil {
-        req.set = req.set.Context(ctx)
+    if sel, ok := req.req.(*tarantool.SelectRequest); !ok {
+        upd := req.req.(*tarantool.UpdateRequest);
+        req.req = upd.Context(ctx)
     } else {
-        req.get = req.get.Context(ctx)
+        req.req = sel.Context(ctx)
     }
 
     return req
@@ -110,42 +99,22 @@ func (req *Request) Context(ctx context.Context) *Request {
 
 // Code returns IPROTO code for the session settings request.
 func (req *Request) Code() int32 {
-    req.assertBuiltWithNew()
-
-    if req.set != nil {
-        return req.set.Code()
-    }
-    return req.get.Code()
+    return req.req.Code()
 }
 
 // Body fills an encoder with the session settings request body.
 func (req *Request) Body(res tarantool.SchemaResolver, enc *encoder) error {
-    req.assertBuiltWithNew()
-
-    if req.set != nil {
-        return req.set.Body(res, enc)
-    }
-    return req.get.Body(res, enc)
+    return req.req.Body(res, enc)
 }
 
 // Ctx returns a context of the session settings request.
 func (req *Request) Ctx() context.Context {
-    req.assertBuiltWithNew()
-
-    if req.set != nil {
-        return req.set.Ctx()
-    }
-    return req.get.Ctx()
+    return req.req.Ctx()
 }
 
 // Async returns is the session settings request expects a response.
 func (req *Request) Async() bool {
-    req.assertBuiltWithNew()
-
-    if req.set != nil {
-        return req.set.Async()
-    }
-    return req.get.Async()
+    return req.req.Async()
 }
 
 // NewErrorMarshalingEnabledSetRequest creates a request to
@@ -272,7 +241,7 @@ func NewSQLVDBEDebugGetRequest() *Request {
 // current session settings in tuple format.
 func NewSessionSettingsGetRequest() *Request {
     return &Request{
-        get: tarantool.NewSelectRequest(sessionSettingsSpace).
+        req: tarantool.NewSelectRequest(sessionSettingsSpace).
             Limit(selectAllLimit),
     }
 }
```

Reply:
I don't see any way to to it properly, because there seems to be no way to declare contextRequest since select and update implement different Context methods: 
```
	*tarantool.SelectRequest does not implement contextRequest (wrong type for Context method)
		have Context(ctx context.Context) *tarantool.SelectRequest
		want Context(ctx context.Context)

	*tarantool.UpdateRequest does not implement contextRequest (wrong type for Context method)
		have Context(ctx context.Context) *tarantool.UpdateRequest
		want Context(ctx context.Context)
```

So in the end I've just decided to implement two separate types of requests: `settings.SetRequest` and `settings.GetRequest`. It seems adequate, the only one drawback is copypaste (which seems like it could be removed with generics later).